File: apps/sepuling/src/main/java/smecalculus/bezmen/core/SepulkaServiceImpl.java
package smecalculus.bezmen.core;

import static java.util.UUID.randomUUID;

import java.time.LocalDateTime;
import java.util.Collections;
import java.util.List;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import smecalculus.bezmen.core.ClientSide.PreviewRequest;
import smecalculus.bezmen.core.ClientSide.PreviewResponse;
import smecalculus.bezmen.core.ClientSide.RegistrationRequest;
import smecalculus.bezmen.core.ClientSide.RegistrationResponse;
import smecalculus.bezmen.storage.SepulkaDao;

@RequiredArgsConstructor
public class SepulkaServiceImpl implements SepulkaService {

    @NonNull
    private SepulkaSliceMapper mapper;

    @NonNull
    private SepulkaDao dao;

    @Override
    public RegistrationResponse register(RegistrationRequest request) {
        var now = LocalDateTime.now();
        var sepulkaCreated = mapper.toServer(request)
                .internalId(randomUUID())
                .revision(0)
                .createdAt(now)
                .updatedAt(now)
                .build();
        var sepulkaSaved = dao.add(sepulkaCreated);
        return mapper.toClient(sepulkaSaved).build();
    }

    @Override
    public List<PreviewResponse> view(PreviewRequest request) {
        return Collections.emptyList();
    }
}

File: apps/sepuling-kotlin/src/main/kotlin/smecalculus/bezmen/core/SepulkaServiceImpl.kt
package smecalculus.bezmen.core

import smecalculus.bezmen.core.MessageDm.PreviewRequest
import smecalculus.bezmen.core.MessageDm.PreviewResponse
import smecalculus.bezmen.storage.SepulkaDao
import java.time.LocalDateTime
import java.util.UUID

class SepulkaServiceImpl(
    private val mapper: SepulkaMapper,
    private val dao: SepulkaDao,
) : SepulkaService {
    override fun register(request: MessageDm.RegistrationRequest): MessageDm.RegistrationResponse {
        val now = LocalDateTime.now()
        val sepulkaCreated =
            mapper.toState(request)
                .internalId(UUID.randomUUID())
                .revision(0)
                .createdAt(now)
                .updatedAt(now)
                .build()
        val sepulkaSaved = dao.add(sepulkaCreated)
        return mapper.toMessage(sepulkaSaved).build()
    }

    override fun view(request: PreviewRequest): List<PreviewResponse> {
        return listOf()
    }
}
